### ISSUE TITLE
spotbugs reports IT_NO_SUCH_ELEMENT for infinite iterator

### ISSUE THREAD

**Issue by hgschmie** [spotbugs/spotbugs#3501]
consider this code:

```java
class ConstantIterator<T> implements Iterator<T> {
    private final T value;

    private ConstantIterator(T value) {
        this.value = value;
    }

    @Override
    public boolean hasNext() {
        return true;
    }

    @Override
    public T next() {
        return value;
    }
}
```

This is a fine iterator. However, spotbugs reports

`[ERROR] Medium: ConstantIterator.next() cannot throw NoSuchElementException [ConstantIterator] At ConstantIterator.java:[line 17] IT_NO_SUCH_ELEMENT`

throwing "NoSuchElementException" is a strictly optional operation in an iterator (the javadoc state "[NoSuchElementException](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/NoSuchElementException.html) - if the iteration has no more elements". This iteration always has more elements.

Spotbugs should not report an error here.

**Comment by gtoison** [comment]
This looks like #1549 but even clearer that the iterator does not need to throw `NoSuchElementException`.
The detector currently does not attempt to find these cases and there are indeed false positives.
In theory the class is not final so it would be possible to override `hasNext()` and break the contract.

This unfortunately falls in the category of false positives that are obvious for human but rather complicated to fix in SpotBugs

**Comment by gtoison** [comment]
Thanks for reporting the issue, the case when hasNext() unconditionally returns true should be fixed by #3548
